src/utf8.hh

```diff
@@ -258,6 +259,18 @@ Iterator character_start(Iterator it, const Sentinel& begin) noexcept
     return it;
 }
 
+// returns an optional iterator to the first byte of the previous character
+// or no value if it is at begin
+template<typename Iterator, typename Sentinel>
+static Optional<Codepoint> prev_codepoint(Iterator it, const Sentinel& begin) noexcept
+{
+    if (it <= begin)
+        return {};
+    auto itlast = utf8::character_start(it - 1, begin);
+    return utf8::read_codepoint(itlast, it);
```

Comment:
no need for utf8:: here as we already are in that namespace. Also, you can just do

```
return codepoint(character_start(it -1, begin), it);
```

Reply:
@mawww Thank you for the patience, I should have seen these things.